### PR TITLE
[WIP] Add AsyncTasks

### PR DESCRIPTION
**Pull request by abhik**
Experimental change to luigi worker to support async tasks. 

Regular luigi tasks have a `run` method that, for async tasks, submits to k8s/etc and blocks until completion. Each task in run in a separate process and this leads to increased mem usage.

`AsyncTask`s, instead, have `submit` and `check_complete` methods. The worker calls `submit`, keeps track of submitted async tasks and periodically calls `check_complete`. All of this happens in the main worker process. After this change (and related change in `sawyer`) we no longer need `KubernetesWait` tasks that repeatedly fail. The luigi dashboard correctly shows async tasks (like `KubernetesRun`) as running rather than pending.

This change is tested on simple Launch tasks and I'm currently testing on large CNA pipelines. The exact method names like `submit` and `check_complete` are still subject to change.

### CODE REVIEW

File: luigi/worker.py [hunk 1]
```diff
@@ -1170,19 +1263,37 @@ def run(self):
         self._add_worker()
 
         while True:
+            # 1. Scheduler doesn't make distinction between sync and async tasks
+            # 2. We only run self.worker_processes concurrent sync tasks
```
Comment (TMaYaD) [inline luigi/worker.py]: Irrespective of async/sync status we should be executing only self.worker_processes number of tasks. `AsyncTask#process` and `AsyncTask#check_status` should be executed in worker_process context.

File: luigi/worker.py [hunk 2]
```diff
@@ -984,14 +1007,84 @@ def _get_work(self):
             worker_state=r.get('worker_state', WORKER_STATE_ACTIVE),
         )
 
+    @ignore_exceptions
+    def _run_async_task(self, task):
```
[the remark this answers is not in the document]
Comment (abhik) [reply luigi/worker.py]: Because TaskProcess calls `run` rather than `submit`/`check`. Merging both into one method wouldn't make sense so we'll just end up moving this code into `TaskProcess` but as new methods. Not sure what benefit that provides.

File: luigi/worker.py [hunk 3]
```diff
@@ -984,14 +1007,84 @@ def _get_work(self):
...
+            missing = [dep.task_id for dep in task.deps() if not dep.complete()]
+            if missing:
+                deps = 'dependency' if len(missing) == 1 else 'dependencies'
+                logger.error('Unfulfilled %s at run time: %s' % (deps, ', '.join(missing)))
+                return
+
+        try:
+            task.trigger_event(Event.START, task)
+            task._start_time = time.time()
+            task.pre_submit()
```
[the remark this answers is not in the document]
Comment (abhik) [reply luigi/worker.py]: See the PR in sawyer. This is used to enter into the context manager that changes targets to temp paths. Doesn't fit the general usecase for events.

File: luigi/worker.py [hunk 4]
```diff
@@ -984,14 +1007,84 @@ def _get_work(self):
...
+            task.submit()
+        except Exception as ex:
+            try:
+                task.post_submit(ex)
```
[the remark this answers is not in the document]
Comment (abhik) [reply luigi/worker.py]: No. post_submit replaces the exit method of the context manager used within run. We need to call post_submit when (1) submission fails and (2) when task completes. There's no submission-failure event in luigi.

Luigi and sawyer don't use events to implement temporary paths. I think it doesn't make sense here for the same reasons.

File: luigi/worker.py [hunk 5]
```diff
@@ -984,14 +1007,84 @@ def _get_work(self):
             worker_state=r.get('worker_state', WORKER_STATE_ACTIVE),
         )
 
+    @ignore_exceptions
```
[the remark this answers is not in the document]
Comment (abhik) [reply luigi/worker.py]: This is temporary because I wanted to log exceptions but not fail the worker. I'll remove before final merge.

File: luigi/worker.py (hunk 1 above)
Comment (abhik) [reply luigi/worker.py]: > Irrespective of async/sync status we should be executing only self.worker_processes number of tasks.

I disagree. The whole point of async tasks is to run many tasks without incurring the memory cost of creating blocking processes. That's why we created KubernetesRun (submit) KubernetesWait (check).

If we set `worker_processes` to `N`, the scheduler won't send work when `N` tasks are already "running". So, to submit many k8s jobs, we'd have to set `worker_processes` to a large number. But, it's then possible for the scheduler to send that many non-async jobs and in that case, we would create many processes and probably get OOMKilled (or have to request a lot of memory beforehand).

File: luigi/worker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TMaYaD) [reply luigi/worker.py]: Exactly the point. The only difference is weather we call submit or run. And the fork in code path should happen where we call run, ie, from within task process. This will keep things DRY and keeps submit calls in sync with run tasks as expected.

File: luigi/worker.py (hunk 1 above)
Comment (TMaYaD) [reply luigi/worker.py]: I agree that "running" tasks shouldn't be accounted for in number of worker processes. But I said "executing" tasks, ie, async task during "submit"/"check status" phase should be treated same as sync task in run phase. It wouldn't be accounted for only when it's sleeping/waiting for next status check.

It may not be the case for kubernetes run but assume some async task with intensive/long running submit method. One would expect it to be restricted by  worker processes count.

File: luigi/worker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (abhik) [reply luigi/worker.py]: There are other differences:

1. TaskProcess runs in new process. We want to run async tasks in main worker process and not incur the time/memory costs. 
2. Sync tasks can yield other tasks as dynamic dependencies. Async tasks cannot.

File: luigi/worker.py (hunk 1 above)
Comment (abhik) [reply luigi/worker.py]: That would mean running submit/check in a separate process and counting that as a "task", even though it's not a luigi task as tracked by scheduler. That seems unnecessarily confusing. Note that the worker uses the number of running tasks to decide when to block on the result queue. So, this also means that the submit/check process will have to use the regular task result queue to communicate with the main worker process. The listener on that queue will then need to decide, for each queue message, whether the result should be propagated to the scheduler or not.

> It may not be the case for kubernetes run but assume some async task with intensive/long running submit method. 

I'm fine, for now, with this change being limited to just our use case. We haven't been able to merge upstream changes in months and merging this change upstream won't change that. I'd rather not generalize beyond our use case until we have more experience with this in prod.

Comment (abhik) [reply luigi/worker.py]: Even if we decide to run submit/check in a separate process, it shouldn't count as a "task" in terms of `self.worker_processes ` and shouldn't use TaskProcess (which is for luigi tasks) or the existing result queue. 

There should only ever be one process submitting/checking all async tasks and if needed, its memory requirements can be taken into account when requesting k8s mem.

File: luigi/worker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TMaYaD) [reply luigi/worker.py]: Put simply, I don't agree with either assumption. What I do agree with is that given those assumptions, this is a decent approach.
But I don't. :D

**Comment by TMaYaD** [comment]
To summarise the discussion so far, We are working under these assumptions:

1. `AsyncTask`'s `submit` and `check_status` must be lite-weight.
  - As an extension, they should be fine with being handled in single process/main process.
2. **Most concerning** Sync tasks can yield other tasks as dynamic dependencies. Async tasks cannot.
3. We are designing for our own use-case and not concerned about upstream for now.
  - pre and post submit are specifically built for sawyer and won't need consistency with events

File: luigi/worker.py [hunk 6]
```diff
@@ -1191,7 +1302,9 @@ def run(self):
                         break
                 else:
                     self._idle_since = None
-                    self._handle_next_task()
+                    self._check_async_tasks()
+                    if len(self._running_tasks):
```
Comment (TMaYaD) [inline luigi/worker.py]: Why is this if check needed. Is this related to our change?

Comment (abhik) [reply luigi/worker.py]: It's the logic that was there before: if there are running tasks, handle next one. The if statements had to change to support async running tasks.

**Comment by abhik** [comment]
> To summarise the discussion so far

1. submit/check
    - Can pass `_request_timeout` to k8s API calls to ensure they don't block for long. I'll add that in sawyer PR.
    - Running `submit` in separate process is easy. Running `check_complete` is trickier because you have to ensure that only one process is checking completeness for a given task. Two processes checking k8s for same task would lead to hairy races. Note that `worker._handle_next_task` (closest analog of `check_complete`) which checks the results of tasks running in other processes also runs in the main worker process.

2. Async tasks can't yield other tasks. We can rewrite such that `submit` can yield other tasks before submitting the k8s job but we don't have such a use case. And, since we're defining a new interface (submit/check rather than run), we're specifying what the new task class can and can't do. AsyncTasks are supposed to be lightweight tasks where the worker simply submits and checks and the actual work is done elsewhere. Allowing `submit` to yield other tasks now means that it can be heavyweight. Note that if you want to yield some tasks and submit, you can always split that into two tasks.

3. pre/post-submit. We might be able to cleanly implement these with existing events. I'll take a look.

**Comment by TMaYaD** [comment]
I'm working on changes towards moving submit and check complete into worker processes. I'll send out a PR later today.
